contribute-to-tidb/review-a-pr-stage.md

```diff
@@ -0,0 +1,233 @@
+
+Sometimes, making choices can be tricky.
+
+Said that there is a special optimization that can improve the overall performance by 30%. However, it introduces a totally different code path and every feature later on has to consider this code path specially, which means a huge maintenance burden. Shall we accept or reject it?
+
+Said that there is a critical bug and a PR is read to fix it, but the change is risky to introduce other kind of bugs. Shall we accept or reject it?
+
+The answer is always “it depends”. Software design is more like a kind of art than technology. It is about aesthetics and your taste of the code. Trade-offs are always there and there is no perfect solution sometimes. Still, there is a general standard (from [@shenli](https://github.com/shenli), maybe we can call it "better than before?"):
+
+* It is making something better, although not a perfect solution.
+* It can mitigate the pain of the user.
+* There is no bug introduced by the change itself.
+
+Then the code can be accepted.
+
+Of course, code review should have some minimal requirements.
+
+For a bug fix PR, if it comes with no test case, the reviewer should ask the author to add tests.
+
+For a performance improvement PR, if no benchmark result is provided, the reviewer should ask the author to attach a benchmark result.
+
+## What to look for in a code review
+
+### Design
+
+The most important thing to cover in a review is the overall design.
+
+Read the description of the PR, and if it's not easy to understand, ask the author to improve it. You may consider questions like:
+
+* Does the change integrate well with the rest of our system?
+* Is it a good time now to add this functionality?
+
+[TiDB Internals forum](https://internals.tidb.io/) is a place for discussion. For pull requests that are created without prior consensus, try to seek consensus there.
+
+### Functionality
+
+* Does this PR do what the developer intend to do?
+* Is what the developer intend to do good for the users with this implementation?
+
+Basically, PR authors are expected to test their PRs well-enough before code reviews. However, as a reviewer you should still think about edge cases, especially for problems about concurrency.
+
+You can always validate the PR if you want. It's hard to understand how some changes will impact a user when you're just reading the code, such as a UI change. You can ask the PR author to give you a demo of the functionality if it's too inconvenient to try it yourself.
+
+### Logic
+
+In the case that concurrent programming involved, it's particularly important to think about the logic in code review. These sorts of issues are very hard to detect by just running the code and usually need somebody to think carefully.
+
+The human brain is just like a special machine, it is good at checking a small piece of logic before the states become too complex. Complexity is the biggest challenge of software.
```

Comment:
I think you can safely delete this paragraph. Remember, your reader's time is limited.

Reply:
Updated.